libmcount/dynamic.c

```diff
@@ -537,13 +537,22 @@ static int do_dynamic_update(struct symtabs *symtabs, char *patch_funcs,
 	}
 
 	while (!list_empty(&patterns)) {
-		struct func_patt_list *pl;
+		struct module_patt_list *ml;
+
+		ml = list_first_entry(&patterns, struct module_patt_list, list);
+		while (!list_empty(&ml->func_patt)) {
+			struct func_patt_list *pl;
 
-		pl = list_first_entry(&patterns, struct func_patt_list, list);
+			pl = list_first_entry(&ml->func_patt, struct func_patt_list, list);
+
+			list_del(&pl->list);
+			free(pl->module);
+			free(pl);
+		}
 
-		list_del(&pl->list);
-		free(pl->module);
-		free(pl);
+		list_del(&ml->list);
+		free(ml->module);
+		free(ml);
```

Comment:
In general, I think it's better to put the release code along with the allocation code.  So we can easily find where they are and make sure not to create any leaks.

Also please consider adding an unit test for new functionalities.  It's also good to check memory errors with ASAN.